testbed/correctness/utils.go

```diff
@@ -0,0 +1,163 @@
+package correctness
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"go.opentelemetry.io/collector/testbed/testbed"
+)
+
+// CreateConfigYaml creates a yaml config for an otel collector given a testbed sender, testbed receiver, any
+// processors, and a pipeline type. A collector created from the resulting yaml string should be able to talk
+// the specified sender and receiver.
+func CreateConfigYaml(
+	sender testbed.DataSender,
+	receiver testbed.DataReceiver,
+	processors map[string]string,
+	pipelineType string,
+) string {
+
+	// Prepare extra processor config section and comma-separated list of extra processor
+	// names to use in corresponding "processors" settings.
+	processorsSections := ""
+	processorsList := ""
+	if len(processors) > 0 {
+		first := true
+		for name, cfg := range processors {
+			processorsSections += cfg + "\n"
+			if !first {
+				processorsList += ","
+			}
+			processorsList += name
+			first = false
+		}
+	}
+
+	format := `
+receivers:%v
+exporters:%v
+processors:
+  %s
+
+extensions:
+
+service:
+  extensions:
+  pipelines:
+    %s:
+      receivers: [%v]
+      processors: [%s]
+      exporters: [%v]
+`
+
+	return fmt.Sprintf(
+		format,
+		sender.GenConfigYAMLStr(),
+		receiver.GenConfigYAMLStr(),
+		processorsSections,
+		pipelineType,
+		sender.ProtocolName(),
+		processorsList,
+		receiver.ProtocolName(),
+	)
+}
+
+// PipelineDef holds the information necessary to run a single testbed configuration.
+type PipelineDef struct {
```

Comment:
Does this type need to be public?

Reply:
A test in the metrics package will need to iterate over a slice of `PipelineDef`s and access `PipelineDef`'s fields, but to my surprise, if I make the type unexported, things still work. So I'll make the change.

Reply:
But making it non-exported causes a lint failure
```
testbed/correctness/utils.go:81:51: exported func LoadPictOutputPipelineDefs returns unexported type []go.opentelemetry.io/collector/testbed/correctness.pipelineDef, which can be annoying to use (golint)
func LoadPictOutputPipelineDefs(fileName string) ([]pipelineDef, error) {
```
so I'll revert that change.

Reply:
Then how is the user of the exported function consume that?

Reply:
I don't know but it compiled and ran. :-) Moot point now since I reverted the change but I thought it was interesting.